File: Packages/RallyUI/Sources/RallyUI/Components/LoadingShimmer.swift
import SwiftUI
import RallyCore

/// Skeleton placeholder view that matches common component shapes and displays
/// a shimmer animation while content is loading.
///
/// Usage:
/// ```swift
/// LoadingShimmer(shape: .card)
/// LoadingShimmer(shape: .row, count: 3)
/// ```
public struct LoadingShimmer: View {

    // MARK: - Shape

    /// Predefined skeleton shapes matching Rally UI component layouts.
    public enum Shape {
        /// Full-width card skeleton (matches RallyCard / ActivationCard).
        case card
        /// Horizontal row skeleton (matches RewardRow / LeaderboardRow).
        case row
        /// Circular badge skeleton (matches PointsBadge).
        case circle(diameter: CGFloat = 72)
        /// Banner skeleton (matches SchoolHeader).
        case banner
        /// Custom rectangular skeleton.
        case rectangle(width: CGFloat? = nil, height: CGFloat)
    }

    // MARK: - Properties

    private let shape: Shape
    private let count: Int

    @State private var phase: CGFloat = 0

    // MARK: - Init

    /// Creates a loading shimmer placeholder.
    /// - Parameters:
    ///   - shape: The skeleton shape to render (default `.card`).
    ///   - count: Number of skeleton items to stack (default `1`).
    public init(shape: Shape = .card, count: Int = 1) {
        self.shape = shape
        self.count = max(1, count)
    }

    // MARK: - Body

    public var body: some View {
        VStack(spacing: RallySpacing.smMd) {
            ForEach(0..<count, id: \.self) { index in
                shimmerContent
                    .animation(
                        .linear(duration: 1.5)
                        .repeatForever(autoreverses: false)
                        .delay(Double(index) * 0.15),
                        value: phase
                    )
            }
        }
        .onAppear {
            phase = 1
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Loading content")
        .accessibilityAddTraits(.updatesFrequently)
    }

    // MARK: - Shape Rendering

    @ViewBuilder
    private var shimmerContent: some View {
        switch shape {
        case .card:
            cardSkeleton
        case .row:
            rowSkeleton
        case .circle(let diameter):
            circleSkeleton(diameter: diameter)
        case .banner:
            bannerSkeleton
        case .rectangle(let width, let height):
            rectangleSkeleton(width: width, height: height)
        }
    }

    // MARK: - Card Skeleton

    private var cardSkeleton: some View {
        VStack(alignment: .leading, spacing: RallySpacing.smMd) {
            // Icon placeholder
            shimmerRect(width: 40, height: 40, radius: RadiusToken.small)

            // Title line
            shimmerRect(height: 18, radius: 4)
                .frame(maxWidth: .infinity)
                .padding(.trailing, RallySpacing.xxxl)

            // Body lines
            shimmerRect(height: 12, radius: 4)
                .frame(maxWidth: .infinity)

            shimmerRect(height: 12, radius: 4)
                .frame(maxWidth: .infinity)
                .padding(.trailing, RallySpacing.xl)

            Spacer(minLength: RallySpacing.sm)

            // Bottom row
            HStack {
                shimmerRect(width: 80, height: 14, radius: 4)
                Spacer()
                shimmerRect(width: 60, height: 14, radius: 4)
            }
        }
        .padding(RallySpacing.md)
        .frame(minHeight: 160)
        .background(
            RoundedRectangle(cornerRadius: RadiusToken.card, style: .continuous)
                .fill(RallyColors.navyMid)
        )
    }

    // MARK: - Row Skeleton

    private var rowSkeleton: some View {
        HStack(spacing: RallySpacing.smMd) {
            // Image placeholder
            shimmerRect(width: 72, height: 72, radius: RadiusToken.small)

            VStack(alignment: .leading, spacing: RallySpacing.sm) {
                shimmerRect(height: 16, radius: 4)
                    .frame(maxWidth: 180)

                shimmerRect(height: 12, radius: 4)
                    .frame(maxWidth: 120)

                shimmerRect(height: 14, radius: 4)
                    .frame(maxWidth: 80)
            }

            Spacer()

            shimmerRect(width: 64, height: 32, radius: RadiusToken.button)
        }
        .padding(RallySpacing.smMd)
        .background(
            RoundedRectangle(cornerRadius: RadiusToken.card, style: .continuous)
                .fill(RallyColors.navyMid)
        )
    }

    // MARK: - Circle Skeleton

    private func circleSkeleton(diameter: CGFloat) -> some View {
        VStack(spacing: RallySpacing.xs) {
            Circle()
                .fill(shimmerGradient)
                .frame(width: diameter, height: diameter)

            shimmerRect(width: 32, height: 10, radius: 4)
        }
    }

    // MARK: - Banner Skeleton

    private var bannerSkeleton: some View {
        ZStack(alignment: .bottomLeading) {
            shimmerRect(height: 180, radius: RadiusToken.card)
                .frame(maxWidth: .infinity)

            HStack(spacing: RallySpacing.smMd) {
                Circle()
                    .fill(Color.white.opacity(0.1))
                    .frame(width: 56, height: 56)

                VStack(alignment: .leading, spacing: RallySpacing.sm) {
                    shimmerRect(width: 160, height: 20, radius: 4)
                        .opacity(0.5)
                    shimmerRect(width: 100, height: 14, radius: 4)
                        .opacity(0.5)
                }
            }
            .padding(RallySpacing.md)
        }
    }

    // MARK: - Rectangle Skeleton

    private func rectangleSkeleton(width: CGFloat?, height: CGFloat) -> some View {
        shimmerRect(width: width, height: height, radius: RadiusToken.small)
            .frame(maxWidth: width == nil ? .infinity : nil)
    }

    // MARK: - Shimmer Building Blocks

    private func shimmerRect(
        width: CGFloat? = nil,
        height: CGFloat,
        radius: CGFloat = 4
    ) -> some View {
        RoundedRectangle(cornerRadius: radius, style: .continuous)
            .fill(shimmerGradient)
            .frame(width: width, height: height)
    }

    private var shimmerGradient: some ShapeStyle {
        LinearGradient(
            colors: [
                RallyColors.gray.opacity(0.15),
                RallyColors.gray.opacity(0.25),
                RallyColors.gray.opacity(0.15)
            ],
            startPoint: UnitPoint(x: phase - 1, y: 0.5),
            endPoint: UnitPoint(x: phase, y: 0.5)
        )
    }
}

// MARK: - Preview

#Preview("Loading Shimmers") {
    ScrollView {
        VStack(spacing: 24) {
            Text("Card Skeleton")
                .font(RallyTypography.caption)
                .foregroundStyle(RallyColors.gray)
            LoadingShimmer(shape: .card)

            Text("Row Skeletons")
                .font(RallyTypography.caption)
                .foregroundStyle(RallyColors.gray)
            LoadingShimmer(shape: .row, count: 3)

            Text("Circle Skeleton")
                .font(RallyTypography.caption)
                .foregroundStyle(RallyColors.gray)
            LoadingShimmer(shape: .circle())

            Text("Banner Skeleton")
                .font(RallyTypography.caption)
                .foregroundStyle(RallyColors.gray)
            LoadingShimmer(shape: .banner)

            Text("Custom Rectangle")
                .font(RallyTypography.caption)
                .foregroundStyle(RallyColors.gray)
            LoadingShimmer(shape: .rectangle(height: 44))
        }
        .padding()
    }
    .background(RallyColors.navy)
}

File: Android/ui/src/main/java/com/vanwagner/rally/ui/components/LoadingShimmer.kt
package com.vanwagner.rally.ui.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.vanwagner.rally.ui.theme.RallyColors
import com.vanwagner.rally.ui.theme.RallyRadius
import com.vanwagner.rally.ui.theme.RallySpacing
import com.vanwagner.rally.ui.theme.RallyTypography

// ---------------------------------------------------------------------------
// LoadingShimmer
// Skeleton placeholder with an infinite shimmer animation.
// Predefined shapes: Card, Row, Circle, Banner, Rectangle.
// ---------------------------------------------------------------------------

/** Predefined skeleton shapes matching Rally UI component layouts. */
sealed class ShimmerShape {
    /** Full-width card skeleton (matches RallyCard / ActivationCard). */
    data object Card : ShimmerShape()

    /** Horizontal row skeleton (matches RewardRow / LeaderboardRow). */
    data object Row : ShimmerShape()

    /** Circular badge skeleton (matches PointsBadge). */
    data class Circle(val diameter: Dp = 72.dp) : ShimmerShape()

    /** Banner skeleton (matches SchoolHeader). */
    data object Banner : ShimmerShape()

    /** Custom rectangular skeleton. */
    data class Rectangle(val width: Dp? = null, val height: Dp) : ShimmerShape()
}

/**
 * Loading shimmer placeholder.
 *
 * @param modifier Optional [Modifier].
 * @param shape    The skeleton shape to render (default [ShimmerShape.Card]).
 * @param count    Number of skeleton items to stack (default 1).
 */
@Composable
fun LoadingShimmer(
    modifier: Modifier = Modifier,
    shape: ShimmerShape = ShimmerShape.Card,
    count: Int = 1,
) {
    val infiniteTransition = rememberInfiniteTransition(label = "shimmer")
    val shimmerOffset by infiniteTransition.animateFloat(
        initialValue = -1f,
        targetValue = 2f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 1500, easing = LinearEasing),
            repeatMode = RepeatMode.Restart,
        ),
        label = "shimmer_offset",
    )

    Column(
        modifier = modifier
            .semantics { contentDescription = "Loading content" },
        verticalArrangement = Arrangement.spacedBy(RallySpacing.smMd),
    ) {
        repeat(count.coerceAtLeast(1)) {
            ShimmerContent(shape = shape, offset = shimmerOffset)
        }
    }
}

@Composable
private fun ShimmerContent(shape: ShimmerShape, offset: Float) {
    when (shape) {
        is ShimmerShape.Card      -> CardSkeleton(offset)
        is ShimmerShape.Row       -> RowSkeleton(offset)
        is ShimmerShape.Circle    -> CircleSkeleton(offset, shape.diameter)
        is ShimmerShape.Banner    -> BannerSkeleton(offset)
        is ShimmerShape.Rectangle -> RectangleSkeleton(offset, shape.width, shape.height)
    }
}

// -- Shimmer brush helper --

@Composable
private fun shimmerBrush(offset: Float): Brush {
    val colors = listOf(
        RallyColors.Gray.copy(alpha = 0.15f),
        RallyColors.Gray.copy(alpha = 0.25f),
        RallyColors.Gray.copy(alpha = 0.15f),
    )
    return Brush.linearGradient(
        colors = colors,
        start = Offset(x = offset * 1000f, y = 0f),
        end = Offset(x = (offset + 1f) * 1000f, y = 0f),
    )
}

@Composable
private fun ShimmerRect(
    offset: Float,
    modifier: Modifier = Modifier,
    width: Dp? = null,
    height: Dp = 16.dp,
    radius: Dp = 4.dp,
) {
    Box(
        modifier = modifier
            .then(if (width != null) Modifier.width(width) else Modifier)
            .height(height)
            .clip(RoundedCornerShape(radius))
            .background(shimmerBrush(offset)),
    )
}

// -- Card Skeleton --

@Composable
private fun CardSkeleton(offset: Float) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(RallyRadius.card))
            .background(RallyColors.NavyMid)
            .padding(RallySpacing.md)
            .height(160.dp),
        verticalArrangement = Arrangement.SpaceBetween,
    ) {
        // Icon placeholder
        ShimmerRect(offset, width = 40.dp, height = 40.dp, radius = RallyRadius.small)

        // Title line
        ShimmerRect(
            offset,
            modifier = Modifier
                .fillMaxWidth()
                .padding(end = RallySpacing.xxxl),
            height = 18.dp,
        )

        // Body lines
        ShimmerRect(offset, modifier = Modifier.fillMaxWidth(), height = 12.dp)
        ShimmerRect(
            offset,
            modifier = Modifier
                .fillMaxWidth()
                .padding(end = RallySpacing.xl),
            height = 12.dp,
        )

        // Bottom row
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            ShimmerRect(offset, width = 80.dp, height = 14.dp)
            ShimmerRect(offset, width = 60.dp, height = 14.dp)
        }
    }
}

// -- Row Skeleton --

@Composable
private fun RowSkeleton(offset: Float) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(RallyRadius.card))
            .background(RallyColors.NavyMid)
            .padding(RallySpacing.smMd),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(RallySpacing.smMd),
    ) {
        // Image placeholder
        ShimmerRect(offset, width = 72.dp, height = 72.dp, radius = RallyRadius.small)

        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(RallySpacing.sm),
        ) {
            ShimmerRect(offset, width = 180.dp, height = 16.dp)
            ShimmerRect(offset, width = 120.dp, height = 12.dp)
            ShimmerRect(offset, width = 80.dp, height = 14.dp)
        }

        ShimmerRect(offset, width = 64.dp, height = 32.dp, radius = RallyRadius.button)
    }
}

// -- Circle Skeleton --

@Composable
private fun CircleSkeleton(offset: Float, diameter: Dp) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(RallySpacing.xs),
    ) {
        Box(
            modifier = Modifier
                .size(diameter)
                .clip(CircleShape)
                .background(shimmerBrush(offset)),
        )
        ShimmerRect(offset, width = 32.dp, height = 10.dp)
    }
}

// -- Banner Skeleton --

@Composable
private fun BannerSkeleton(offset: Float) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(180.dp)
            .clip(RoundedCornerShape(RallyRadius.card))
            .background(shimmerBrush(offset)),
        contentAlignment = Alignment.BottomStart,
    ) {
        Row(
            modifier = Modifier.padding(RallySpacing.md),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(RallySpacing.smMd),
        ) {
            Box(
                modifier = Modifier
                    .size(56.dp)
                    .clip(CircleShape)
                    .background(Color.White.copy(alpha = 0.1f)),
            )
            Column(verticalArrangement = Arrangement.spacedBy(RallySpacing.sm)) {
                ShimmerRect(offset, width = 160.dp, height = 20.dp)
                ShimmerRect(offset, width = 100.dp, height = 14.dp)
            }
        }
    }
}

// -- Rectangle Skeleton --

@Composable
private fun RectangleSkeleton(offset: Float, width: Dp?, height: Dp) {
    ShimmerRect(
        offset = offset,
        width = width,
        height = height,
        radius = RallyRadius.small,
        modifier = if (width == null) Modifier.fillMaxWidth() else Modifier,
    )
}

// MARK: - Previews

@Preview(name = "LoadingShimmer - All Shapes", showBackground = true, backgroundColor = 0xFF131B2E)
@Composable
private fun LoadingShimmerPreview() {
    Surface(color = RallyColors.Navy) {
        Column(
            modifier = Modifier
                .padding(RallySpacing.md)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(RallySpacing.lg),
        ) {
            Text("Card Skeleton", style = RallyTypography.caption, color = RallyColors.Gray)
            LoadingShimmer(shape = ShimmerShape.Card)

            Text("Row Skeletons", style = RallyTypography.caption, color = RallyColors.Gray)
            LoadingShimmer(shape = ShimmerShape.Row, count = 3)

            Text("Circle Skeleton", style = RallyTypography.caption, color = RallyColors.Gray)
            LoadingShimmer(shape = ShimmerShape.Circle())

            Text("Banner Skeleton", style = RallyTypography.caption, color = RallyColors.Gray)
            LoadingShimmer(shape = ShimmerShape.Banner)

            Text("Custom Rectangle", style = RallyTypography.caption, color = RallyColors.Gray)
            LoadingShimmer(shape = ShimmerShape.Rectangle(height = 44.dp))
        }
    }
}
